zebrad: Added a command to dump routes in support bundle

rgirada opened the pull request:
Description:
This command already added to support_bundle_commands.conf file but not defined.
So, Added this show command("show ip zebra route dump") which dumps all routes
with detailed information including next hops, flags, status ..etc.
This route information will be dump to a file when support bundle generated.
This helps for  better dubugging and added this command as a hidden command.

Signed-off-by: Rajesh Girada <rgirada@vmware.com>

qlyoung commented:
@rgirada can you file an issue for that bug then?

donaldsharp commented:
Are we really suggesting that it's better to add a *whole* new command and associated code instead of fixing a bit of already installed code in the code base as well as not actually reporting a memory leak?  Color me unimpressed.

rgirada commented:
@donaldsharp , @qlyoung 
This is not a memory leak with the command , I meant a high memory usage with json command  causing the system to  problematic state. 
When “show ip route json “ is used  with  2 million routes, it is taking more 30 mins to get the response. Sometimes no response and  system goes to hung state. At this time, We have observed , it is occupying more than 87% of system memory to frame the json o/p. Due to this, sometimes watchfrr restart triggers and other daemons killed due to SIGINT signal.

But with the new command, it is taking less than 2 minutes to dump all the 2 million routes.

rgirada commented on doc/user/zebra.rst:
```diff
@@ -1010,6 +1010,13 @@ zebra Terminal Mode Commands
+.. clicmd:: show <ip|ipv6> zebra route dump [<vrf> VRFNAME]
+
+   It dumps all the routes from RIB with detailed information including
+   internal flags,status ..etc.This is defined as hidden command and
```
done

rgirada commented on zebra/zebra_vty.c:
```diff
@@ -1784,6 +1787,209 @@ DEFPY (show_route_summary,
+              SHOW_STR
+              IP_STR
+              IP6_STR
+              "zebra daemon\n"
```
done

rgirada commented on zebra/zebra_vty.c:
```diff
@@ -1784,6 +1787,209 @@ DEFPY (show_route_summary,
+
+	char buf[SRCDEST2STR_BUFFER];
+
+	vty_out(vty, "   Nexthop%u :\n", num);
```
yes white spaces for formate the o/p . Done.

rgirada commented on zebra/zebra_vty.c:
```diff
@@ -1784,6 +1787,209 @@ DEFPY (show_route_summary,
+	case NEXTHOP_TYPE_IPV4_IFINDEX:
+		vty_out(vty, "      ip address: %s\n",
+			inet_ntoa(nexthop->gate.ipv4));
+		vty_out(vty, "      afi : ipv4\n");
```
done.

rgirada commented on zebra/zebra_vty.c:
```diff
@@ -1784,6 +1787,209 @@ DEFPY (show_route_summary,
+		vty_out(vty, "      unreachable : True\n");
+		switch (nexthop->bh_type) {
+		case BLACKHOLE_REJECT:
+			vty_out(vty, "      reject : True\n");
```
Modified the o/p.

rgirada commented on zebra/zebra_vty.c:
```diff
@@ -1784,6 +1787,209 @@ DEFPY (show_route_summary,
+	char buf[SRCDEST2STR_BUFFER];
+
+	vty_out(vty, "   Nexthop%u :\n", num);
+	vty_out(vty, "      type : %d\n", nexthop->type);
```
done.

ton31337 commented:
where we are here?

rgirada commented:
> where we are here?

Hi @ton31337 ,
Thanks for concerning about this PR.
I have finished addressing all review comments and I requested for review . It is ready to merge .

ton31337 commented:
Please rebase before reviewing again.

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/37c4b29141c0c4bba4dd8c284b8335cb/raw/13a4cca28493dc97a223fdeb0f284f062b4c4eab/cr_6317_1600775344.diff | git apply
```

</p>

<p>

```diff
diff --git a/zebra/zebra_vty.c b/zebra/zebra_vty.c
index 721ad7f58..928349912 100644
--- a/zebra/zebra_vty.c
+++ b/zebra/zebra_vty.c
@@ -1937,16 +1937,12 @@ DEFPY (show_route_summary,
 	return CMD_SUCCESS;
 }
 
-DEFUN_HIDDEN (show_route_zebra_dump,
-              show_route_zebra_dump_cmd,
-              "show <ip|ipv6> zebra route dump [<vrf> VRFNAME]",
-              SHOW_STR
-              IP_STR
-              IP6_STR
-              "Zebra daemon\n"
-              "Routing table\n"
-              "All information\n"
-              VRF_CMD_HELP_STR)
+DEFUN_HIDDEN(show_route_zebra_dump, show_route_zebra_dump_cmd,
+	     "show <ip|ipv6> zebra route dump [<vrf> VRFNAME]",
+	     SHOW_STR IP_STR IP6_STR
+	     "Zebra daemon\n"
+	     "Routing table\n"
+	     "All information\n" VRF_CMD_HELP_STR)
 {
 	afi_t afi;
 	struct route_table *table;

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

donaldsharp commented:
@polychaeta autoclose in 2 weeks

polychaeta reviewed (dismissed):
Thanks for your contribution to FRR!

* `sprintf, strcat, strcpy, inet_ntoa` are banned; please use `snprintf, strlcat, strlcpy, inet_ntop`

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

rgirada commented:
Hi @ton31337 ,

Sorry, delay from myside.
I have rebased it , could you help please re-review ?

polychaeta commented:
This issue will no longer be automatically closed.

rgirada commented:
Ci:rerun

rgirada commented:
ci:rerun

qlyoung commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
+	case NEXTHOP_TYPE_IPV4:
+	case NEXTHOP_TYPE_IPV4_IFINDEX:
+		vty_out(vty, "      ip address: %s\n",
+			inet_ntoa(nexthop->gate.ipv4));
```
`inet_ntoa` is a banned function, use `inet_ntop`

rgirada replied:
addressed

qlyoung commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
+	switch (nexthop->type) {
+	case NEXTHOP_TYPE_IPV4:
+	case NEXTHOP_TYPE_IPV4_IFINDEX:
+		vty_out(vty, "      ip address: %s\n",
```
Why is there a space before `:` in the above output but not here? Please be consistent. I prefer you delete the extraneous spaces on lines 2108 - 2110.

rgirada replied:
Made it consistant

qlyoung commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
+	case NEXTHOP_TYPE_IPV4_IFINDEX:
+		vty_out(vty, "      ip address: %s\n",
+			inet_ntoa(nexthop->gate.ipv4));
+		vty_out(vty, "      afi : ipv4\n");
```
Again - be consistent. Please remove the space between `afi` and `:`

rgirada replied:
Done

qlyoung commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
+		if (nexthop->ifindex) {
+			vty_out(vty, "      interfaceIndex : %d\n",
+				nexthop->ifindex);
+			vty_out(vty, "      interfaceName : %s\n",
```
As before with `:`. Also, why are these camelCased? This isn't machine friendly output, so I assume it's for humans - in which case it's clearer to write "Interface index" and "Interface name"

rgirada replied:
Done

qlyoung commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
+		if (nexthop->src.ipv4.s_addr) {
+			if (inet_ntop(AF_INET, &nexthop->src.ipv4, buf,
+				      sizeof(buf)))
+				vty_out(vty, "      source : %s\n", buf);
```
As before with `:`

rgirada replied:
Done

qlyoung commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
+		}
+
+		if (nexthop->src.ipv4.s_addr) {
+			if (inet_ntop(AF_INET, &nexthop->src.ipv4, buf,
```
You can collapse this condition into the outer one

rgirada replied:
done

qlyoung commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
+		break;
+	case NEXTHOP_TYPE_IPV6:
+	case NEXTHOP_TYPE_IPV6_IFINDEX:
+		vty_out(vty, "      ip : %s\n",
```
`:` again. Please fix it on all the remaining statements, I'm not going to comment on the rest.

rgirada replied:
Done

qlyoung commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
+		if (nexthop->ifindex) {
+			vty_out(vty, "      interfaceIndex : %d\n",
+				nexthop->ifindex);
+			vty_out(vty, "      interfaceName : %s\n",
```
As before

rgirada replied:
Addressed

qlyoung commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
+		}
+		break;
+	case NEXTHOP_TYPE_IFINDEX:
+		vty_out(vty, "      Nexthop is ifindex(directlyConnected).\n");
```
Please clean up this message. "Nexthop is an interface (directly connected)".

rgirada replied:
Done

qlyoung commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
+		break;
+	case NEXTHOP_TYPE_BLACKHOLE:
+		vty_out(vty,
+			"      Nexthop is Blackhole interface(unreachable).\n");
```
"Nexthop type is blackhole"

rgirada replied:
Done

qlyoung commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
+			vty_out(vty, "   protocol : %s\n",
+				zebra_route_string(re->type));
+			vty_out(vty, "   instance : %u\n", re->instance);
+			vty_out(vty, "   vrfId : %u\n", re->vrf_id);
```
"VRF ID", though I'm pretty sure you could print the VRF name here.

rgirada replied:
Added vhf name as well , done

qlyoung commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
+			tm = gmtime(&uptime);
+
+			if (uptime < ONE_DAY_SECOND)
+				snprintf(time, 20, "%02d:%02d:%02d",
```
Do not hard code buffer sizes. Use `sizeof(time)`. This comment applies to subsequent uses as well.

rgirada replied:
done

qlyoung commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
+
+			uptime = monotime(&tv);
+			uptime -= re->uptime;
+			tm = gmtime(&uptime);
```
I prefer you use `gmtime_r` as this function isn't thread safe.

rgirada replied:
addressed

qlyoung commented:
@rgirada Please address the blocking review comments. The reason noone reviewed this for so long is because it had obvious issues that were never fixed. You have to respond to review comments or your patch is just going to sit here.

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/fa7e95203533f087f0b71de1434372f5/raw/85c2df7a8affc694cb270e8c0e6029e0af7eac48/cr_6317_1619721542.diff | git apply
```

</p>

<p>

```diff
diff --git a/zebra/zebra_vty.c b/zebra/zebra_vty.c
index 52826e795..f99263946 100644
--- a/zebra/zebra_vty.c
+++ b/zebra/zebra_vty.c
@@ -2060,16 +2060,12 @@ DEFPY (show_route_summary,
 	return CMD_SUCCESS;
 }
 
-DEFUN_HIDDEN (show_route_zebra_dump,
-              show_route_zebra_dump_cmd,
-              "show <ip|ipv6> zebra route dump [vrf VRFNAME]",
-              SHOW_STR
-              IP_STR
-              IP6_STR
-              "Zebra daemon\n"
-              "Routing table\n"
-              "All information\n"
-              VRF_CMD_HELP_STR)
+DEFUN_HIDDEN(show_route_zebra_dump, show_route_zebra_dump_cmd,
+	     "show <ip|ipv6> zebra route dump [vrf VRFNAME]",
+	     SHOW_STR IP_STR IP6_STR
+	     "Zebra daemon\n"
+	     "Routing table\n"
+	     "All information\n" VRF_CMD_HELP_STR)
 {
 	afi_t afi = AFI_IP;
 	struct route_table *table;

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/406e7c248ea67725b8722a539ee54aed/raw/83394897333c0e8189770b05c2b0d2714e841e9d/cr_6317_1619785259.diff | git apply
```

</p>

<p>

```diff
diff --git a/zebra/zebra_vty.c b/zebra/zebra_vty.c
index a3097e6c0..7ffdebcd6 100644
--- a/zebra/zebra_vty.c
+++ b/zebra/zebra_vty.c
@@ -2060,16 +2060,12 @@ DEFPY (show_route_summary,
 	return CMD_SUCCESS;
 }
 
-DEFUN_HIDDEN (show_route_zebra_dump,
-              show_route_zebra_dump_cmd,
-              "show <ip|ipv6> zebra route dump [vrf VRFNAME]",
-              SHOW_STR
-              IP_STR
-              IP6_STR
-              "Zebra daemon\n"
-              "Routing table\n"
-              "All information\n"
-              VRF_CMD_HELP_STR)
+DEFUN_HIDDEN(show_route_zebra_dump, show_route_zebra_dump_cmd,
+	     "show <ip|ipv6> zebra route dump [vrf VRFNAME]",
+	     SHOW_STR IP_STR IP6_STR
+	     "Zebra daemon\n"
+	     "Routing table\n"
+	     "All information\n" VRF_CMD_HELP_STR)
 {
 	afi_t afi = AFI_IP;
 	struct route_table *table;

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

rgirada commented:
> @rgirada Please address the blocking review comments. The reason noone reviewed this for so long is because it had obvious issues that were never fixed. You have to respond to review comments or your patch is just going to sit here.

Hi @qlyoung , 
It is a delay from my side , thanks for reviewing.
I have addressed all your suggested changes .

rgirada commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
 
+DEFUN_HIDDEN (show_route_zebra_dump,
+              show_route_zebra_dump_cmd,
+              "show <ip|ipv6> zebra route dump [<vrf> VRFNAME]",
```
addressed

rgirada commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
+	if (argv_find(argv, argc, "ip", &idx))
+		afi = AFI_IP;
+
+	if (argv_find(argv, argc, "ipv4", &idx)
```
addressed

rgirada commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
+
+	if (argv_find(argv, argc, "ipv4", &idx)
+	    || argv_find(argv, argc, "ipv6", &idx))
+		afi = strmatch(argv[idx]->text, "ipv6") ? AFI_IP6 : AFI_IP;
```
addressed

rgirada commented on zebra/zebra_vty.c:
```diff
@@ -2041,6 +2044,213 @@ DEFPY (show_route_summary,
+	} else {
+		vrf_id_t vrf_id = VRF_DEFAULT;
+
+		if (vrf_name)
```
addressed

qlyoung commented on doc/user/zebra.rst:
```diff
@@ -1058,6 +1058,13 @@ zebra Terminal Mode Commands
+.. clicmd:: show <ip|ipv6> zebra route dump [<vrf> VRFNAME]
+
+   It dumps all the routes from RIB with detailed information including
+   internal flags,status ..etc. This is defined as hidden command and
```
Please fix the basic English issues here

"internal flags, status etc."
"This is defined as a hidden command"

rgirada replied:
Addressed .

qlyoung commented on doc/user/zebra.rst:
```diff
@@ -1058,6 +1058,13 @@ zebra Terminal Mode Commands
+
+   It dumps all the routes from RIB with detailed information including
+   internal flags,status ..etc. This is defined as hidden command and
+   this is added to support_bundle_commands.conf command.
```
I don't think anyone reading this will necessarily know what "support_bundle_commands.conf" is. Either clarify it or remove this bit

rgirada replied:
Addressed.

qlyoung reviewed (commented):
Fix the nits and it's good

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/265edef249dc41b085281f0ae2a932ac/raw/c007d08013f2d665ac2ab6816b296f3d0c8094fc/cr_6317_1620294072.diff | git apply
```

</p>

<p>

```diff
diff --git a/zebra/zebra_vty.c b/zebra/zebra_vty.c
index a3097e6c0e..7ffdebcd6e 100644
--- a/zebra/zebra_vty.c
+++ b/zebra/zebra_vty.c
@@ -2060,16 +2060,12 @@ DEFPY (show_route_summary,
 	return CMD_SUCCESS;
 }
 
-DEFUN_HIDDEN (show_route_zebra_dump,
-              show_route_zebra_dump_cmd,
-              "show <ip|ipv6> zebra route dump [vrf VRFNAME]",
-              SHOW_STR
-              IP_STR
-              IP6_STR
-              "Zebra daemon\n"
-              "Routing table\n"
-              "All information\n"
-              VRF_CMD_HELP_STR)
+DEFUN_HIDDEN(show_route_zebra_dump, show_route_zebra_dump_cmd,
+	     "show <ip|ipv6> zebra route dump [vrf VRFNAME]",
+	     SHOW_STR IP_STR IP6_STR
+	     "Zebra daemon\n"
+	     "Routing table\n"
+	     "All information\n" VRF_CMD_HELP_STR)
 {
 	afi_t afi = AFI_IP;
 	struct route_table *table;

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

rgirada commented:
Hi @qlyoung , Hi @ton31337 
I have addressed all the suggested comments , could you please rereview and approve it  ?

rzalamena commented on doc/user/zebra.rst:
```diff
@@ -1058,6 +1058,12 @@ zebra Terminal Mode Commands
    nexthop groups that do have an afi. [type] allows you to filter those
    only coming from a specific NHG type (protocol).
 
+.. index:: show <ip|ipv6> zebra route dump [<vrf> VRFNAME]
```
Instead of creating a new command, how about augmenting `show ip route`?

Here is a suggestion:
`show ip route details`

qlyoung replied:
How strongly do you feel about this? This PR has been open for a year, I'd like to take care of it and move on.

rzalamena commented on zebra/zebra_vty.c:
```diff
@@ -2057,6 +2060,207 @@ DEFPY (show_route_summary,
+		RB_FOREACH (vrf, vrf_name_head, &vrfs_by_name) {
+			zvrf = vrf->info;
+			if ((zvrf == NULL)
+			    || (zvrf->table[afi][SAFI_UNICAST] == NULL))
```
Could you please also dump multicast routes?

qlyoung replied:
That would be useful for me but again, it's a year old, multicast routes can be added later.

rzalamena reviewed (commented):
Regaring JSON being slow, have you tried removing the JSON pretty output flag? The pretty printed JSON output adds a lot of newlines and spaces.

Unfortunately using the `json_` API we use double the memory: once for JSON data structure representation and again for stringnification of that. One possible optimization (if we want to go this way) is to generate the JSON output directly instead of using the API to save memory.

qlyoung reviewed (approved):
fine
